Improve the featured image UI when it cannot retrieve the image file and data

afercia opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Fixes https://github.com/WordPress/gutenberg/issues/66005

## What?
<!-- In a few words, what is the PR actually doing? -->
The Post featured image UI doesn't take into account the edge case when there is an attachment ID stored in the database but no actual image file and no image meta data. For example, this may happen when importing posts via the WordPress Importer and skipping to import the attachments. It may happen in other cases as well as migrating the database etc.

In these cases, the UI only shows an empty button. See https://github.com/WordPress/gutenberg/issues/66005#issuecomment-2470490093

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The UI should always be clearly operable and provide feedback on the current state including errors or missing attachment / data.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Introduces logic to determine whether the request for the featured image media data has completed.
- When the request has completed and there is an attachment ID but no media data:
  - Shows a paragraph with an informative message instead of the image preview.
  - Shows the Replace and Remove buttons, with a secondary variant and `isDestructive` styling for the remove button.
- When the image is correctly retrieved and rendered, there are no visual changes.
- Added bonus: shows the  loading Spinner also when the request to retrieve the image data is pending. Previously, the Spinner was only shown after dragging an image into the UI, when the upload was pending.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Rename this file to have an `xml` extension: 
[featured image broken.txt](https://github.com/user-attachments/files/17718377/featured.image.broken.txt)
- Import the file via the WordPress Importer under Tools > Import > WordPress importer.
- Do not check the `Download and import file attachments` checkbox.
- At this point you should have a new post with title `Broken featured image`.
- Edit: an easier way to reproduce, as @mamaduka suggested, is by using WP CLI. You need to set a post with a featured image ID that points to a non-existing attachment e.g. `wp post meta update your_post_id_here _thumbnail_id 999999`.
- Edit the post.
- Observe the Featured image UI shows the Spinner while attempting to retrieve the image data.
- When the request completes, observe the UI shows an informative message in a paragraph.
- Observe the Replace and Remove buttons are visible by default (no need to hover or focus) so that the UI is clearly operable.
- Try to remove / replace the image and check everything works as expected.
- Observe that when the image is correctly set there are no visual changes and the buttons only appear on hover or focus, as expected based on the current design.


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

This is the UI when the attachment ID exists but no other data exist:

![Screenshot 2024-11-12 at 15 23 49](https://github.com/user-attachments/assets/9a2b9a0a-e147-424c-bb04-c0061d69a985)

I have no strong opinions on whether the informative message in the paragraph should have some specific styling e.g. like a warning. I'd lean towards keeping it simple. Cc @WordPress/gutenberg-design 

This is the UI with an image correctly set. No changes here:

![Screenshot 2024-11-12 at 15 29 30](https://github.com/user-attachments/assets/04a9bb70-359b-43f8-9781-cb3e5ab9067c)


It's worth noting the case covered here is different from another edge case:
- The featured image ID and all its media data exist in the database.
- The image is later deleted from the filesystem but all its data are still available.

In this case, the UI will show a 'broken image' as browsers do by default, with the alternative text / fallback text displayed. Screenshot:

![Screenshot 2024-11-12 at 15 24 24](https://github.com/user-attachments/assets/09a2d5b6-6bb2-40e4-8f1b-f6a73c993c5a)

That is the current behavior of this UI. Hovering or focusing will show the Replace and Remove buttons but they don't look that great on a white background. I'd tend to think the styling for this case should be improved but it's out of the scope of this PR.

afercia committed:
Improve the featured image UI when cannot retrieve the image file and data.

Mamaduka commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -150,6 +155,25 @@ function PostFeaturedImage( {
 		);
 	}
 
+	const { isRequestingFeaturedImageMedia } = useSelect(
+		( select ) => {
+			const _isRequestingFeaturedImageMedia =
+				!! featuredImageId &&
+				! select( coreStore ).hasFinishedResolution( 'getMedia', [
+					featuredImageId,
+					{ context: 'view' },
+				] );
+
+			return {
+				isRequestingFeaturedImageMedia: _isRequestingFeaturedImageMedia,
+			};
+		},
+		[ featuredImageId ]
+	);
```
Let's select the resolution status in `withSelect` below. We're already selecting media there, and grouping these makes sense. Plus, we'll use existing store subscriptions.

Mamaduka commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -150,6 +155,25 @@ function PostFeaturedImage( {
+		[ featuredImageId ]
+	);
+
+	const noImageData =
+		! isRequestingFeaturedImageMedia && !! featuredImageId && ! media;
```
Nitpick: Using `isMissingMedia` or something similar might be more descriptive here.

Mamaduka commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -174,52 +198,80 @@ function PostFeaturedImage( {
+												) }
+											/>
+										) }
+										{ ( isLoading ||
+											isRequestingFeaturedImageMedia ) && (
+											<Spinner />
+										) }
```
Do we need to add the `isRequestingFeaturedImageMedia` condition here? The values selected via `hasFinishedResolution` rarely work for loading states.

afercia replied:
Actually, this is the bit that makes the Spinner work also when normally selecting an image from the Media library. Previously, the spinner appeared only _when dragging_ an image to the drop zone. The existing `isLoading` name is a bit misleading. At first I thought it was to check the actual 'loading' of an image but no, it is true only when dragging. A better name would have been `isUploading` or something like that.

Mamaduka replied:
The block editor only displays spinners when uploading an image; We shouldn't display a snipper when image data is fetched from the server.

P.S. Snippers should always be a last resort; they make sense as uploading indicators but should be avoided as loading indicators whenever possible, IMO>

afercia replied:
> The block editor only displays spinners when uploading an image

Hum, I'm not sure that would be a good design principle. Not even sure it's a documented guideline? I see Spinners used for many things e.g. loading Pages, Patterns, etc. also Spinners used in documentation examples for other things like delete buttons etc.
Anyways, without the Spinner, this UI would only show an _empty_ button which isn't a good pattern IMHO. Screenshot:

![Screenshot 2024-11-14 at 09 32 30](https://github.com/user-attachments/assets/89874dfc-6726-4b46-addc-6c0e2ef87e66)

Mamaduka commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -174,52 +198,80 @@ function PostFeaturedImage( {
+											{
+												'editor-post-featured-image__actions-missing-image':
+													noImageData,
+												'editor-post-featured-image__actions-is-requesting-image':
+													isRequestingFeaturedImageMedia,
```
What does this class do here?

afercia replied:
What class? 🙂

Mamaduka replied:
The `editor-post-featured-image__actions-is-requesting-image`.

afercia replied:
That class is needed to prevent the Replace and Remove buttons to appear when hovering _while the request is in progress_. Without it, the buttons would appear when hovering. Screenshot:

![Screenshot 2024-11-14 at 09 41 03](https://github.com/user-attachments/assets/b8e1dd86-1abf-43fe-84d5-9084bfb245cc)

afercia committed:
Address CR feedback.

afercia committed:
Use a Notice instead of a paragraph.

afercia commented:
@Mamaduka thanks for your review. 
I addressed part of your points. Also, I've opted to use a Notice instead of a paragraph. Added bonus: the Notice automatically triggers a `speak` message for screen reader users. Screenshot:

![Screenshot 2024-11-14 at 09 24 36](https://github.com/user-attachments/assets/124b3ee7-4880-4e71-9eee-476584711953)

Rebased.

afercia commented:
> Thank you, @afercia!
> 
> I think [this point](https://github.com/WordPress/gutenberg/pull/66936#discussion_r1841732739) is still valid. For consistency, we shouldn't display a loading state when fetching image data from the server.

Thank you @Mamaduka. I kindly disagree. Showing an empty button is less than ideal. The UI state needs to be communicated. It is fetching the image data, to me this is no different from fetching Pages or Patterns and similar states, where a Spinner is used to communicate the retrieval of data.

Mamaduka commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -250,6 +297,12 @@ const applyWithSelect = withSelect( ( select ) => {
 	const featuredImageId = getEditedPostAttribute( 'featured_media' );
+	const isRequestingFeaturedImageMedia =
+		!! featuredImageId &&
+		! select( coreStore ).hasFinishedResolution( 'getMedia', [
```
We're already extracting the `coreStore` selectors above. Let's extract `hasFinishedResolution` there as well.

Mamaduka commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -230,6 +271,12 @@ function PostFeaturedImage( {
 												onRemoveImage();
 												toggleRef.current.focus();
```
This line triggers an error when trying to remove the missing featured image because the toggle button is no longer rendered. We'll need to fix the focus return on removal logic.

## Screenshot
![CleanShot 2024-11-14 at 17 52 52](https://github.com/user-attachments/assets/6cf1d16e-b666-4aa8-aa2d-a8108c93b6e3)

afercia replied:
Good catch 👍🏻

Mamaduka commented:
Okay, I experimented with the proposed snipper a little more, and it doesn't seem out of place. Generally, we should avoid using spinners as loading indicators when possible. Nobody likes seeing dozens of spinners on a page when they open it.

@afercia, I found a [bug with focus return logic](https://github.com/WordPress/gutenberg/pull/66936#discussion_r1842272468), which is currently a main blocker here.

P.S. I've also found an easier way to replicate the bug. WP CLI command:

```
wp post meta update <post-id> _thumbnail_id 999999
```

afercia committed:
Address more CR feedback.

afercia commented:
@mamadka thanks for your thorough review. The focus bug should be fixed now.

Mamaduka reviewed (approved):
> Also, I really wish we didn't have to resort to using `useEffect` and `setTimeout` for focus management, but that's a more general problem with React. It's not a blocker here.

Yeah, I know. I tried to make the toggle alwasy rendered and toggling its visibility in order to avoid a setTimeout. It didn't work anyways.

afercia commented:
Rebased.
